File: V2/files (2)/domain-pipeline-schema/src/fragments.ts
/**
 * Fragment Schemas — Sub-Agent Output Types
 *
 * Sub-agents produce fragments. Fragments are smaller, easier to validate,
 * and don't carry the full pipeline envelope. The assembler merges fragments
 * into the final Node output.
 *
 * Key difference from final schemas:
 * - No EventIDs, GapIDs, ObstacleIDs — the assembler assigns these
 * - Predecessor/successor links use event NAMES, not IDs
 * - No pipeline envelope — fragments are intermediate, not stored long-term
 *   (though they CAN be stored for debugging/audit)
 */

import { z } from "zod";
import { AggCode, CodeRef, HotSpot, Severity, StateType, TransitionAnnotation, ResolutionType } from "./primitives.js";

// ── Node 1 Fragment ─────────────────────────────────────────────────────────
// One per scan area. Sub-agent reads code, extracts events, links by name.

export const EventFragment = z.object({
  /** PascalCase past tense. This is the dedup key. */
  name: z
    .string()
    .regex(/^[A-Z][a-zA-Z0-9]+$/, "PascalCase event name"),
  trigger: z.string().min(1),
  location: CodeRef,
  /** Link by name, not ID. Assembler resolves to EventIDs. */
  predecessorNames: z.array(z.string()),
  /** Link by name, not ID. Assembler resolves to EventIDs. */
  successorNames: z.array(z.string()),
  stateChange: z.string().min(1),
  evidence: z.string().min(10, "Evidence must contain actual code observation"),
  hotSpots: z.array(HotSpot),
});
export type EventFragment = z.infer<typeof EventFragment>;

export const Node1Fragment = z.object({
  areaCode: AggCode,
  areaName: z.string().min(1),
  filesScanned: z.array(z.string().min(1)).min(1),
  filesSkipped: z.array(z.string()),
  events: z.array(EventFragment),
});
export type Node1Fragment = z.infer<typeof Node1Fragment>;

// ── Node 2 Fragment ─────────────────────────────────────────────────────────
// One per aggregate. Sub-agent discovers states (ReAct) and maps transitions.
// Gap IDs are provisional (just sequential numbers) — assembler prefixes them.

export const StateFragment = z.object({
  name: z.string().min(1),
  type: StateType,
  representation: z.string().min(1),
  location: CodeRef,
  evidence: z.string().min(10),
});

export const TransitionFragment = z.object({
  source: z.string().min(1),         // state name
  target: z.string().min(1),         // state name
  eventName: z.string().min(1),      // event NAME, not ID — assembler resolves
  guard: z.string().nullable(),
  sideEffects: z.array(z.string()),
  annotation: TransitionAnnotation.nullable(),
});

export const GapFragment = z.object({
  /** Sequential number only. Assembler prefixes with AggCode. */
  seq: z.number().int().min(1),
  severity: Severity,
  description: z.string().min(1),
  codeLocation: CodeRef,
});

export const Node2Fragment = z.object({
  aggCode: AggCode,
  aggName: z.string().min(1),
  rootEntity: z.string().min(1),
  keyFiles: z.array(CodeRef).min(1),
  states: z.array(StateFragment).min(1),
  transitions: z.array(TransitionFragment),
  gaps: z.array(GapFragment),
  trivialLifecycle: z.boolean(),
});
export type Node2Fragment = z.infer<typeof Node2Fragment>;

// ── Node 3 Fragment ─────────────────────────────────────────────────────────
// One per aggregate. Forward obstacle analysis results.
// Obstacle IDs are provisional sequential numbers — assembler assigns O-N{seq}.

export const ObstacleFragment = z.object({
  /** Sequential number. Assembler assigns global O-N{seq}. */
  seq: z.number().int().min(1),
  /** Goal IDs (these are already assigned by the orchestrator). */
  violatesGoals: z.array(z.string()).min(1),
  evidence: z.string().min(1),
  description: z.string().min(1),
  severity: Severity,
  resolutionType: ResolutionType,
});

export const Node3Fragment = z.object({
  aggCode: AggCode,
  aggName: z.string().min(1),
  newObstacles: z.array(ObstacleFragment),
});
export type Node3Fragment = z.infer<typeof Node3Fragment>;

File: ede/fragments.py
"""
Fragment Schemas -- Sub-Agent Output Types

Sub-agents produce fragments. Fragments are smaller, easier to validate,
and don't carry the full pipeline envelope. The assembler merges fragments
into the final Node output.

Key difference from final schemas:
- No EventIDs, GapIDs, ObstacleIDs -- the assembler assigns these
- Predecessor/successor links use event NAMES, not IDs
- No pipeline envelope -- fragments are intermediate, not stored long-term
  (though they CAN be stored for debugging/audit)
"""

from __future__ import annotations

from typing import Annotated

from pydantic import Field, StringConstraints

from .primitives import (
    AggCode,
    CodeRef,
    EdeBaseModel,
    HotSpot,
    ResolutionType,
    Severity,
    StateType,
    TransitionAnnotation,
)


# -- Node 1 Fragment ----------------------------------------------------------
# One per scan area. Sub-agent reads code, extracts events, links by name.

class EventFragment(EdeBaseModel):
    """Event without assigned ID. Uses event name for linking."""

    name: Annotated[str, StringConstraints(pattern=r"^[A-Z][a-zA-Z0-9]+$")]
    trigger: str = Field(min_length=1)
    location: CodeRef
    predecessor_names: list[str]
    successor_names: list[str]
    state_change: str = Field(min_length=1)
    evidence: str = Field(min_length=10)
    hot_spots: list[HotSpot]


class Node1Fragment(EdeBaseModel):
    area_code: AggCode
    area_name: str = Field(min_length=1)
    files_scanned: list[str] = Field(min_length=1)
    files_skipped: list[str]
    events: list[EventFragment]


# -- Node 2 Fragment ----------------------------------------------------------
# One per aggregate. Sub-agent discovers states and maps transitions.
# Gap IDs are provisional (just sequential numbers) -- assembler prefixes them.

class StateFragment(EdeBaseModel):
    name: str = Field(min_length=1)
    type: StateType
    representation: str = Field(min_length=1)
    location: CodeRef
    evidence: str = Field(min_length=10)


class TransitionFragment(EdeBaseModel):
    source: str = Field(min_length=1)
    target: str = Field(min_length=1)
    event_name: str = Field(min_length=1)  # event NAME, not ID
    guard: str | None
    side_effects: list[str]
    annotation: TransitionAnnotation | None
    evidence: str = Field(min_length=10)
    code_location: CodeRef


class GapFragment(EdeBaseModel):
    """Sequential number only. Assembler prefixes with AggCode."""

    seq: int = Field(ge=1)
    severity: Severity
    description: str = Field(min_length=1)
    code_location: CodeRef


class Node2Fragment(EdeBaseModel):
    agg_code: AggCode
    agg_name: str = Field(min_length=1)
    root_entity: str = Field(min_length=1)
    key_files: list[CodeRef] = Field(min_length=1)
    states: list[StateFragment] = Field(min_length=1)
    transitions: list[TransitionFragment]
    gaps: list[GapFragment]
    trivial_lifecycle: bool


# -- Node 3 Fragment ----------------------------------------------------------
# One per aggregate. Forward obstacle analysis results.

class ObstacleFragment(EdeBaseModel):
    """Sequential number. Assembler assigns global O-N{seq}."""

    seq: int = Field(ge=1)
    violates_goals: list[str] = Field(min_length=1)
    evidence: str = Field(min_length=1)
    description: str = Field(min_length=1)
    severity: Severity
    resolution_type: ResolutionType


class Node3Fragment(EdeBaseModel):
    agg_code: AggCode
    agg_name: str = Field(min_length=1)
    new_obstacles: list[ObstacleFragment]
